File: packages/android-im-ui/src/test/kotlin/com/flare/im/ui/QrEncoderTest.kt
package com.flare.im.ui

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull

/** FR-031: the kit QR encoder against the qrcodegen reference matrices in [QrOracleData]. */
class QrEncoderTest {
    private fun rows(symbol: QrSymbol): List<String> = (0 until symbol.size).map { y ->
        buildString { for (x in 0 until symbol.size) append(if (symbol.isDark(x, y)) '1' else '0') }
    }

    /** FNV-1a 32 over the concatenated row strings (same as the oracle dump). */
    private fun fnv1a32(rows: List<String>): String {
        var hash = 0x811c9dc5.toInt()
        for (row in rows) for (unit in row) hash = (hash xor unit.code) * 0x01000193
        return hash.toUInt().toString(16).padStart(8, '0')
    }

    private fun darkCount(rows: List<String>) = rows.sumOf { row -> row.count { it == '1' } }

    @Test fun matchesReferenceMatricesBitForBit() {
        for (case in QrOracleData.cases) {
            val symbol = assertNotNull(
                QrEncoder.encodeText(case.payload, case.forcedMask.takeIf { it >= 0 }),
                case.id,
            )
            assertEquals(case.version, symbol.version, "${case.id}: smallest version")
            assertEquals(4 * case.version + 17, symbol.size, "${case.id}: size")
            val maskKind = if (case.forcedMask < 0) "automatic mask choice" else "forced mask"
            assertEquals(case.mask, symbol.mask, "${case.id}: $maskKind")
            val actual = rows(symbol)
            for (y in actual.indices) assertEquals(case.rows[y], actual[y], "${case.id}: row $y")
        }
    }

    @Test fun forcedCasesCoverAllEightMasks() {
        assertEquals((0..7).toSet(), QrOracleData.cases.map { it.forcedMask }.filter { it >= 0 }.toSet())
    }

    @Test fun versionSweepAtByteCapacityMatchesReference() {
        assertEquals((1..40).toList(), QrOracleData.sweep.map { it.version })
        for (s in QrOracleData.sweep) {
            val payload = String(CharArray(s.length) { i -> (33 + (i * 7 + s.version * 13) % 94).toChar() })
            val auto = assertNotNull(QrEncoder.encodeText(payload), "v${s.version}")
            assertEquals(s.version, auto.version, "v${s.version}: version")
            assertEquals(s.autoMask, auto.mask, "v${s.version}: automatic mask choice")
            val autoRows = rows(auto)
            assertEquals(s.autoDark, darkCount(autoRows), "v${s.version}: dark modules")
            assertEquals(s.autoHash, fnv1a32(autoRows), "v${s.version}: matrix hash")

            val forced = assertNotNull(QrEncoder.encodeText(payload, s.forcedMask), "v${s.version} forced")
            assertEquals(s.forcedMask, forced.mask)
            val forcedRows = rows(forced)
            assertEquals(s.forcedDark, darkCount(forcedRows), "v${s.version} mask ${s.forcedMask}: dark modules")
            assertEquals(s.forcedHash, fnv1a32(forcedRows), "v${s.version} mask ${s.forcedMask}: matrix hash")
        }
    }

    @Test fun picksTheSmallestVersionAtEveryCapacityBoundary() {
        for (version in 1 until 40) {
            val fits = ByteArray(QrOracleData.capacity[version]) { 0x61 }
            assertEquals(version, QrEncoder.encodeBytes(fits)?.version, "${fits.size} bytes")
            assertEquals(version + 1, QrEncoder.encodeBytes(fits + 0x61.toByte())?.version, "${fits.size + 1} bytes")
        }
    }

    @Test fun rejectsPayloadsLongerThanVersion40Holds() {
        assertNotNull(QrEncoder.encodeBytes(ByteArray(QrOracleData.capacity[40]) { 0x61 }))
        assertNull(QrEncoder.encodeBytes(ByteArray(QrOracleData.capacity[40] + 1) { 0x61 }))
    }
}

File: packages/flutter-im-ui/test/qr_encoder_test.dart
import 'package:flare_im_ui/src/components/qr/flare_qr_encoder.dart';
import 'package:flutter_test/flutter_test.dart';

import 'qr_oracle_data.dart';

List<String> _rows(QrSymbol symbol) => [
  for (var y = 0; y < symbol.size; y++)
    String.fromCharCodes([
      for (var x = 0; x < symbol.size; x++) symbol.isDark(x, y) ? 0x31 : 0x30,
    ]),
];

/// FNV-1a 32 over the concatenated row strings (same as the oracle dump).
String _fnv1a32(List<String> rows) {
  var hash = 0x811c9dc5;
  for (final row in rows) {
    for (final unit in row.codeUnits) {
      hash = (hash ^ unit) & 0xFFFFFFFF;
      hash = (hash * 0x01000193) & 0xFFFFFFFF;
    }
  }
  return hash.toRadixString(16).padLeft(8, '0');
}

int _darkCount(List<String> rows) =>
    rows.fold(0, (sum, row) => sum + '1'.allMatches(row).length);

void main() {
  group('QrEncoder matches the qrcodegen reference bit for bit', () {
    for (final c in qrOracleCases) {
      test(c.id, () {
        final symbol = QrEncoder.encodeText(
          c.payload,
          mask: c.forcedMask < 0 ? null : c.forcedMask,
        );
        expect(symbol, isNotNull);
        expect(symbol!.version, c.version, reason: 'smallest version');
        expect(symbol.size, 4 * c.version + 17);
        expect(
          symbol.mask,
          c.mask,
          reason: c.forcedMask < 0 ? 'automatic mask choice' : 'forced mask',
        );
        final rows = _rows(symbol);
        for (var y = 0; y < rows.length; y++) {
          expect(rows[y], c.rows[y], reason: 'row $y');
        }
      });
    }

    test('forced cases cover all eight mask patterns', () {
      expect(
        {
          for (final c in qrOracleCases)
            if (c.forcedMask >= 0) c.forcedMask,
        },
        {0, 1, 2, 3, 4, 5, 6, 7},
      );
    });
  });

  group('version sweep at byte capacity (versions 1…40)', () {
    test('covers every version', () {
      expect(
        [for (final s in qrOracleSweep) s.version],
        [for (var v = 1; v <= 40; v++) v],
      );
    });

    for (final s in qrOracleSweep) {
      test('version ${s.version}', () {
        final payload = String.fromCharCodes([
          for (var i = 0; i < s.length; i++) 33 + (i * 7 + s.version * 13) % 94,
        ]);
        final auto = QrEncoder.encodeText(payload)!;
        expect(auto.version, s.version);
        expect(auto.mask, s.autoMask);
        final autoRows = _rows(auto);
        expect(_darkCount(autoRows), s.autoDark);
        expect(_fnv1a32(autoRows), s.autoHash);

        final forced = QrEncoder.encodeText(payload, mask: s.forcedMask)!;
        expect(forced.mask, s.forcedMask);
        final forcedRows = _rows(forced);
        expect(_darkCount(forcedRows), s.forcedDark);
        expect(_fnv1a32(forcedRows), s.forcedHash);
      });
    }
  });

  test('picks the smallest version at every capacity boundary', () {
    for (var version = 1; version < 40; version++) {
      final fits = List.filled(qrOracleCapacity[version], 0x61);
      expect(QrEncoder.encodeBytes(fits)!.version, version);
      expect(QrEncoder.encodeBytes([...fits, 0x61])!.version, version + 1);
    }
  });

  test('rejects payloads longer than version 40 holds', () {
    expect(
      QrEncoder.encodeBytes(List.filled(qrOracleCapacity[40], 0x61)),
      isNotNull,
    );
    expect(
      QrEncoder.encodeBytes(List.filled(qrOracleCapacity[40] + 1, 0x61)),
      isNull,
    );
  });
}
